### PR TITLE
notify/fuse: first working implementation

### PR DESCRIPTION
**Pull request by rjeczalik**
This PR adds first working implementation of FUSE-based notifier for a mount subsystem.

Deps: ~#10588 #10590~

There's going to be another PR with cmd/loopfuse - a tool I used to test current implementation. Since rsync-based syncer is not yet integrated, I did not write any integration tests just to not have to rework them after the rsync is in. The integration tests we're going to write separately after gathering feedback from more wide internal testing.

### CODE REVIEW

**Comment by rjeczalik** [comment]
I think something change with github and I'm not longer able to create forward PRs that depend on others - trying to create a PR from koding:loopfuse branch fails with:

```
< HTTP 422
{"message":"Validation Failed","errors":[{"resource":"PullRequest","code":"custom","message":"A pull request already exists for koding:loopfuse."}],"documentation_url":"https://developer.github.com/v3/pulls/#create-a-pull-request"}
Error creating pull request: Unprocessable Entity (HTTP 422)
A pull request already exists for koding:loopfuse.
```

The tool waits at this branch:

https://github.com/koding/koding/tree/loopfuse/go/src/koding/klient/machine/mount/notify/fuse/cmd/loopfuse

I'm going to send a PR after this one is merged.

**Comment by ppknap** [comment]
@rjeczalik could you rebase when #10588 is merged?

**Comment by rjeczalik** [comment]
@ppknap Rebased.

File: go/src/koding/klient/machine/mount/notify/fuse/filesystem.go [hunk 1]
```diff
@@ -0,0 +1,624 @@
...
+	fs.Remote.PromiseDel(path, nd)
+	atomic.StoreUint64(&nd.Entry.Inode, 0)
+
+	path = fs.path(path)
+
+	if err := os.Remove(path); os.IsNotExist(err) {
+		return nil
+	}
+
+	return fs.yield(ctx, path, index.ChangeMetaLocal|index.ChangeMetaRemove)
```
Comment (ppknap) [inline go/src/koding/klient/machine/mount/notify/fuse/filesystem.go]: we don't need to wait here for update

File: go/src/koding/klient/machine/mount/notify/fuse/filesystem.go [hunk 2]
```diff
@@ -0,0 +1,624 @@
...
+}
+
+func (fs *Filesystem) update(ctx stdcontext.Context, f *os.File, nd *index.Node) error {
+	if err := f.Sync(); err != nil {
+		return err
+	}
+
+	updateSize(f, nd)
+
+	return fs.yield(ctx, f.Name(), index.ChangeMetaLocal|index.ChangeMetaUpdate)
```
Comment (ppknap) [inline go/src/koding/klient/machine/mount/notify/fuse/filesystem.go]: no need to wait here

File: go/src/koding/klient/machine/mount/notify/fuse/fuse.go [hunk 3]
```diff
@@ -0,0 +1,481 @@
...
+	path = filepath.Join(path, op.Name)
+
+	op.Entry.Child, err = fs.mkdir(path, op.Mode)
+	if err != nil {
+		return err
+	}
+
+	op.Entry.Attributes = fs.newAttr(op.Mode)
+
+	return fs.yield(ctx, path, index.ChangeMetaAdd|index.ChangeMetaLocal)
```
Comment (ppknap) [inline go/src/koding/klient/machine/mount/notify/fuse/fuse.go]: do we need to wait here?

File: go/src/koding/klient/machine/mount/notify/fuse/fuse.go [hunk 4]
```diff
@@ -0,0 +1,481 @@
...
+	path = filepath.Join(path, op.Name)
+
+	op.Entry.Child, op.Handle, err = fs.mkfile(path, op.Mode)
+	if err != nil {
+		return err
+	}
+
+	op.Entry.Attributes = fs.newAttr(op.Mode)
+
+	return fs.yield(ctx, path, index.ChangeMetaAdd|index.ChangeMetaLocal)
```
Comment (ppknap) [inline go/src/koding/klient/machine/mount/notify/fuse/fuse.go]: no need to wait here.

File: go/src/koding/klient/machine/mount/notify/fuse/fuse.go [hunk 5]
```diff
@@ -0,0 +1,481 @@
...
+	delete(fs.inodes, id)
+	id = fs.add(newPath)
+	atomic.StoreUint64(&entry.Inode, uint64(id))
+	fs.mu.Unlock()
+
+	fs.Remote.PromiseDel(oldPath, oldNd)
+	fs.Remote.PromiseAdd(newPath, entry)
+
+	return nonil(
+		fs.yield(ctx, oldPath, index.ChangeMetaLocal|index.ChangeMetaRemove),
```
Comment (ppknap) [inline go/src/koding/klient/machine/mount/notify/fuse/fuse.go]: no need to wait here
```go
fs.Cache.Commit(index.NewChange(oldpath, index.ChangeMetaRemove))
fs.Cache.Commit(index.NewChange(newpath, index.ChangeMetaAdd))

return nil
```

File: go/src/koding/klient/machine/mount/notify/fuse/fuse.go [hunk 6]
```diff
@@ -0,0 +1,481 @@
...
+		return fuse.EIO
+	}
+
+	path = filepath.Join(path, op.Name)
+
+	if err := fs.rm(ctx, nd, path); err != nil {
+		return err
+	}
+
+	return fs.yield(ctx, path, index.ChangeMetaLocal|index.ChangeMetaRemove)
```
Comment (ppknap) [inline go/src/koding/klient/machine/mount/notify/fuse/fuse.go]: no wait

File: go/src/koding/klient/machine/mount/notify/fuse/fuse.go [hunk 7]
```diff
@@ -0,0 +1,481 @@
...
+		return err
+	}
+
+	if _, err = f.WriteAt(trimRightNull(op.Data), op.Offset); err != nil {
+		return err
+	}
+
+	updateSize(f, nd)
+
+	return fs.yield(ctx, f.Name(), index.ChangeMetaLocal|index.ChangeMetaUpdate)
```
Comment (ppknap) [inline go/src/koding/klient/machine/mount/notify/fuse/fuse.go]: no need to wait

**Review by ppknap — COMMENTED** [review]
FUSE doesn't have to wait for all syncs since it has some files locally available just after the call

File: go/src/koding/klient/machine/mount/notify/fuse/export_test.go [hunk 8]
```diff
@@ -0,0 +1,4 @@
+package fuse
+
+// TrimRightNull exports trimRightNull for testing purposes.
```
Comment (cihangir) [inline go/src/koding/klient/machine/mount/notify/fuse/export_test.go]: 👍

File: go/src/koding/klient/machine/mount/notify/fuse/filesystem.go [hunk 9]
```diff
@@ -0,0 +1,624 @@
...
+	f, err := os.Create(absPath)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	if err := f.Close(); err != nil {
+		return 0, 0, err
+	}
+
+	if f, err = os.OpenFile(absPath, os.O_RDWR, 0644); err != nil {
```
Comment (rjeczalik) [inline go/src/koding/klient/machine/mount/notify/fuse/filesystem.go]: Done, but different way round - removed the second open.

File: go/src/koding/klient/machine/mount/notify/fuse/filesystem.go (hunk 1 above)
Comment (rjeczalik) [inline go/src/koding/klient/machine/mount/notify/fuse/filesystem.go]: Done.

File: go/src/koding/klient/machine/mount/notify/fuse/filesystem.go (hunk 2 above)
Comment (rjeczalik) [inline go/src/koding/klient/machine/mount/notify/fuse/filesystem.go]: Done.

File: go/src/koding/klient/machine/mount/notify/fuse/fuse.go (hunk 3 above)
Comment (rjeczalik) [inline go/src/koding/klient/machine/mount/notify/fuse/fuse.go]: Done.

File: go/src/koding/klient/machine/mount/notify/fuse/fuse.go (hunk 4 above)
Comment (rjeczalik) [inline go/src/koding/klient/machine/mount/notify/fuse/fuse.go]: Done.

File: go/src/koding/klient/machine/mount/notify/fuse/fuse.go (hunk 5 above)
Comment (rjeczalik) [inline go/src/koding/klient/machine/mount/notify/fuse/fuse.go]: Done.

File: go/src/koding/klient/machine/mount/notify/fuse/fuse.go (hunk 6 above)
Comment (rjeczalik) [inline go/src/koding/klient/machine/mount/notify/fuse/fuse.go]: Done.

File: go/src/koding/klient/machine/mount/notify/fuse/fuse.go (hunk 7 above)
Comment (rjeczalik) [inline go/src/koding/klient/machine/mount/notify/fuse/fuse.go]: Done.

**Comment by rjeczalik** [comment]
@ppknap @cihangir All addressed, ptal.

**Review by ppknap — APPROVED** [review]

**Comment by cihangir** [comment]
i haven't checked this PR since it is bigger than our limits, is there any way to decrease the size and add unit tests?

**Comment by rjeczalik** [comment]
@cihangir I can rework it into PRs, each than contain only one method implementation for fuse.FileSystem. It was not possible to do that from the begining, since I needed working fuse implementation to ensure the code od correct. As it was not possible to prove it with unittests, I've created a tool - loopfuse - in a separate PR. I will add also unittests for the methods as well.

File: go/src/koding/klient/fs/statfs_linux.go [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+	var stfs syscall.Statfs_t
+
+	if err := syscall.Statfs(path, &stfs); err != nil {
+		return nil, err
+	}
+
+	di := &DiskInfo{
+		BlockSize:   uint32(stfs.Bsize),
+		BlocksTotal: stfs.Blocks,
+		BlocksFree:  stfs.Bfree,
```
Comment (ppknap) [inline go/src/koding/klient/fs/statfs_linux.go]: Is there a reason why `IOSize` is missing here? I used it on my linux machine.

**Comment by ppknap** [comment]
>  I will add also unittests for the methods as well.

Could it be done in separate PR after this is merged? @cihangir this PR may be hard to split into smaller pieces. If this help: `filesystem.go` is a priv interface for `fuse.go` impl (which implements `jacobsa/go-fuse`). Rest of the files are not directly related. This was my approach to review this (quite large) repo.

It would be very nice to have this merged as soon as possible.

**Comment by cihangir** [comment]
It seems you want this PR in so bad @ppknap :)

> Could it be done in separate PR after this is merged?

historically we dont do that most of the time, something _more_ important comes up and we postpone it to a _later_ time.

I would like to see PRs with their unit tests next to them, so my vote is 👎 for merging. But if you both guys say OK to merge you can go ahead and merge. I have reviewed PRs three times bigger than this so the size doesnt matter i just dont want to break the rules we agreed upon.

File: go/src/koding/klient/fs/statfs_linux.go (hunk 10 above)
Comment (rjeczalik) [inline go/src/koding/klient/fs/statfs_linux.go]: https://github.com/golang/go/blob/master/src/syscall/ztypes_linux_amd64.go#L101-L117

How?

**Comment by rjeczalik** [comment]
@cihangir @ppknap We have two options:

1) have a devel branch for mounts, develop there until we reach a beta version that can be deployed to dev channel and be tested by QA; then later rework it to be production-ready code and send nice PRs
2) merge it now and rework later on master

I always opt for 1) as prevents merging code to master and then forgetting to rework it, which happens way too many times.

I will rework* FUSE after I'm done with 1) kloud 2) kite fixes ("unknown error occurred") 3) kdio. More details in an internal mail that I will send later.

* scout's word

**Comment by ppknap** [comment]
> It seems you want this PR in so bad @ppknap :)

Hah, you discovered my evil plan of merging this. I'm bored of being HTML developer :)

> have a devel branch for mounts,

I spent about one hour fixing >100 merge conflicts in >15 commits. Trying to avoid this.

**Comment by rjeczalik** [comment]
> I spent about one hour fixing >100 merge conflicts in >15 commits. Trying to avoid this.

Becase we're not developing against devel branch.

You create a devel branch, branch off your fork branch, work on a PR, after it's ready and won't be touched again you push it to devel. Then rebase your another feature branch against devel. I did not have excessive merge conflicts following that flow. Unless I happen to edit the same file that my fellow developer.

After everything is ready we slice the devel branch in PRs and send in order as gh PRs against master.

**Comment by cihangir** [comment]
https://github.com/koding/koding/pull/10591#issuecomment-282761694

If you say we will make it better, we could merge it to master as well.

**Comment by ppknap** [comment]
@rjeczalik before we merge, some of `Commit` calls get an absolute path of the change and some get relative. Could you fix this and make all of them relative to cache directory?

**Comment by rjeczalik** [comment]
@ppknap Done in 1a072415f5be8309b3a55989e6990ba60e3e5fc3.

**Comment by ppknap** [comment]
@cihangir @rjeczalik what about merging this? I will tests these changes anyway.

image

**Pull request merged**